server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAlias.java

```diff
@@ -313,11 +398,24 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeString(name);
         out.writeStringCollection(dataStreams);
         out.writeOptionalString(writeDataStream);
-        if (filter != null) {
-            out.writeBoolean(true);
-            filter.writeTo(out);
+        if (out.getVersion().onOrAfter(Version.V_8_7_0)) {
+            out.writeMap(dataStreamToFilterMap, StreamOutput::writeString, (out1, filter) -> filter.writeTo(out1));
         } else {
-            out.writeBoolean(false);
+            if (dataStreamToFilterMap.isEmpty()) {
+                out.writeBoolean(false);
+            } else {
+                /*
+                 * Versions before 8.7 incorrectly only allowed a single filter for all datastreams, and randomly dropped all others. We
+                 * replicate that buggy behavior here if we have to write to an older node because there is no way to send multipole
+                 * filters to an older node.
+                 */
+                logger.info(
+                    "Sending DataStreamAlias {} to a node with version earlier than 8.7.0. Versions before 8.7.0 do not support "
+                        + "separate filters for each DataStream.",
+                    name
+                );
```

Comment:
I don't think we should log for binary serialization. I think the logging we have for XContent parsing is useful (and only executed once), but mixed version nodes could cause this to be spammed, so we should remove this.

Reply:
My fear was that people would be accidentally corrupting their data stream aliases and that spamming would be better than nothing here. Is this not going to be a problem in practice?